### ISSUE TITLE
Inconsistent ICA decomposition between runs on Linux

### ISSUE THREAD

**Issue by notZaki** [ME-ICA/tedana#629]
### Summary

The classification of the ICA decomposition appears to change across runs, as noted in #625. Initially, it looked as if the python version had an impact on the results, which is why the gif below shows the python version. However, further testing found that re-running with the same python version can give different results.
 ![](https://user-images.githubusercontent.com/9019681/101269640-a6bb8180-3736-11eb-851a-f5416ce9e75e.gif)

**Update: 2020/12/09**

The different sources that can cause variations in the decomposition are:

- Hardware
- Operating system
    + Technically, this part isn't nailed down. None of the tests have used tried different OS on the same hardware.
- Python version / software
- ~Using `tedana_cli.tedana_workflow` vs `tedana_cli._main`~
    + This difference was caused because `_main` defaults to using 1 thread, while calling `tedana_workflow` directly results in all threads being used. The number of threads seems to change the decomposition.

---

### Additional Detail

Python versions `3.5` to `3.9` were tested on mac and linux using the three-echo test data:

https://github.com/notZaki/tedana/blob/7cfd19cd6d20d75f52d4f42fe01613beb61bce9d/tedana/tests/test_integration.py#L189-L208

Although the seed was not specified, it should default to `42`.

The number of accepted/rejected/ignored ICA components was printed at the end of the log:
https://github.com/notZaki/tedana/blob/7cfd19cd6d20d75f52d4f42fe01613beb61bce9d/tedana/workflows/tedana.py#L659-L662

The table below was made by collecting the results from [these logs](https://github.com/notZaki/tedana/actions/runs/406654556).
Tests on MacOS gave consistent results, but Linux had outliers on Python 3.6 and 3.7.

| Python | OS    | Accepted | Rejected | Ignored | Total |
|--------|-------|----------|----------|---------|-------|
| 3.5    | Linux | 13       | 23       | 32      | 68    |
| 3.6    | Linux | **12** | **25** |**31** | 68    |
| 3.7    | Linux | **12** | **25** |**31** | 68    |
| 3.8    | Linux | 13       | 23       | 32      | 68    |
| 3.9    | Linux | 13       | 23       | 32      | 68    |
| 3.5    | Mac   | 13       | 23       | 32      | 68    |
| 3.6    | Mac   | 13       | 23       | 32      | 68    |
| 3.7    | Mac   | 13       | 23       | 32      | 68    |
| 3.8    | Mac   | 13       | 23       | 32      | 68    |
| 3.9    | Mac   | 13       | 23       | 32      | 68    |

The same runs were repeated [once](https://github.com/notZaki/tedana/actions/runs/406680591) and then [again](https://github.com/notZaki/tedana/actions/runs/406735097) and [again²](https://github.com/notZaki/tedana/actions/runs/406786495). 
~MacOS remained consistent, but the Linux runs showed further inconsistency:~
- ~Python 3.7 on Linux sometimes returns `13-23-32` [[log]](https://github.com/notZaki/tedana/runs/1513271543?check_suite_focus=true)~
- ~Python 3.8 on Linux sometimes returns `12-25-31` [[log]](https://github.com/notZaki/tedana/runs/1513123731?check_suite_focus=true)~ 
- ~Python 3.9 on Linux sometimes returns `12-25-31` [[log]](https://github.com/notZaki/tedana/runs/1513426203?check_suite_focus=true)~
The variations on the Linux runs were due to hardware differences. 

### Next Steps

- **Question:** Is the output of ICA decomposition supposed to be consistent or are minor variations expected/tolerable?
- The current runs are only displaying for the number of components at the end of the log. Additional details could be checked, such as the total percentage variance explained by the accepted/rejected/ignored components. 
- Check whether the denoised output is substantially different between the `13-23-32` vs `12-25-31` cases.
- ~Repeat the runs a few more times to verify that MacOS remain robust~
- Add Windows to my tests---have to figure out why there's an error while installing dependencies [[log]](https://github.com/notZaki/tedana/runs/1512655180?check_suite_focus=true)

**Comment by jbteves** [comment]
Thanks for opening this @notZaki ! I am a little surprised that you are able to run 3.5 with the latest version of `tedana`, as I believe the bokeh version we require needs Python 3.6+. That said, we'll have to look into this with the test data. I'll open a separate issue about installing `tedana` on Windows.

**Comment by jbteves** [comment]
@all-contributors add @notZaki for bug reporting.

**Comment by jbteves** [comment]
Sorry to tag you so much @notZaki, but...
@all-contributors please add @notZaki for bug reporting and code.

**Comment by eurunuela** [comment]
Hey @notZaki , thank you for opening this issue!

Did those "outliers" converge on the first try? Or was the seed updated to help with convergence? If the seed doesn't change, I wouldn't expect changes in the ICA results. The classification is another thing though.

So, first, I think we should identify the "outliers" to see why we get the different results.

If the ICA components don't match, then the only reason that comes to mind right now is that OpenBlas or whatever FastICA is running on on your Linux behaves differently depending on the Python version. Another option is that the "outliers" are somehow close to the tolerance, but I think this is very unlikely to happen.

I would expect them to match and to see differences in the classification.

**Comment by jbteves** [comment]
Quick Linux question: what version of linux (e.g., Ubuntu 20.04 LTS or CentOS 7), and via containerization (like docker) or running on the native OS?

**Comment by notZaki** [comment]
The table was made using runs on the default setup used by github actions. Based on the logs, their current configuration uses Ubuntu 18.04.5 LTS with [this virtual environment](https://github.com/actions/virtual-environments/blob/ubuntu18/20201129.1/images/linux/Ubuntu1804-README.md)

The gif is from running natively on CentOS 7, where Py 3.6 is using the global libraries that are available to all users, and Py 3.8 was using a conda environment---or maybe it was just 3.8 from anconda and I forgot to activate the environment, will have to repeat that.

**Comment by notZaki** [comment]
@eurunuela I believe they all converged on the first try. 

I printed the seed value before and after the ICA decomposition---assuming this was the right function: https://github.com/notZaki/tedana/commit/2816cfbfce24864ebdba0ac8cefdf6464e1e12bc
The seed remains at `42` while the classification either ends up as `12-25-31` [[log]](https://github.com/notZaki/tedana/runs/1514132383?check_suite_focus=true) or `13-23-32` [[log]](https://github.com/notZaki/tedana/runs/1514181243?check_suite_focus=true) as the number of `accepted-rejected-ignored` ICA components.

If it was purely a software issue, then I wouldn't expect to see any variations on github-actions, but even the same python version can give different results across different instances.

---

I've tried a few more things. The results are shared in the fragments below. I'm not sure how to piece them together yet.

---

**Differences in hardware are part of the problem.** Although github-actions uses the same software, the linux builds can use two different family of CPUs: 

- Intel(R) Xeon(R) CPU **E5**-2673 v4 @ 2.30GHz
- Intel(R) Xeon(R) **Platinum** 8171M or 8272CL CPU @ 2.60GHz

If the E5 is used, then the decomposition always results in `13-23-32`.
If the Platinum is used, then the decomposition always results in `12-25-31`. (Technically not true, see next fragment)
I've checked Python [3.6](https://github.com/notZaki/tedana/actions/runs/407191363), [3.7](https://github.com/notZaki/tedana/actions/runs/407159735), [3.8](https://github.com/notZaki/tedana/actions/runs/407236398) and [3.9](https://github.com/notZaki/tedana/actions/runs/407209845)---the results are always consistent with the CPU.

I don't know what CPU the MacOS builds use on github-actions, but I'm guessing MacOS is more picky about hardware and probably use the same CPU---which would explain why all the MacOS runs gave consistent results in the top-level table.

So to make our results reproducible by others, we would need to ship our code and the CPU. Is there a docker for CPUs? 😬

---

**The choice of using `tedana_cli` or `tedana_workflow` can affect the decomposition.** The table below shows the number of `accepted-rejected-ignored` ICA components and is based off [these logs](https://github.com/notZaki/tedana/actions/runs/407517757). The Platinum and E5 CPU is used by github-actions for Linux builds, while I wasn't able to figure what CPU the MacOS builds use.

| CPU       | CLI      | Workflow |
|-----------|----------|----------|
| Platinum  | 12-25-31 | 13-23-32 |
| E5        | 13-23-32 | 13-23-32 |
| MacOS     | 13-23-32 | 12-25-31 |

Oh, and my system is running Linux on an E5-2667 and behaves identical to MacOS in the above table. 😕 
Maybe this explains how I managed to get different results in the gif despite using the same CPU. Although I don't remember switching between the workflow/cli. 

---

**On github-actions, Python 3.5 on Linux always returns 13-23-31 regardless of CPU or choice of cli/workflow.** [[log]](https://github.com/notZaki/tedana/actions/runs/407540021). Unfortunately, Python 3.5 in not the panacea for all systems---MacOS builds still suffer from the discrepancy in the above table. 

On my system, I get the opposite effect---all runs result in 12-25-31 on Python 3.5 regardless of whether cli/workflow is used. At least it's consistent? 🤷

**Comment by eurunuela** [comment]
> The choice of using tedana_cli or tedana_workflow can affect the decomposition.

This doesn't make any sense. I've just gone through the defaults and both the CLI and the workflow have the same defaults. So, I think it should be run 10/20 times to really verify that the CLI and workflow yield different results. Same for the hardware.

The logs are fine but they only tell us about the number of accepted, rejected, and ignored components. In order to properly understand where the difference in results comes from, we need to look at each ICA component and make sure we get the same components with the different conditions (regardless of their classification). My guess is there are one or two components that are tricky for tedana to classify and we need to see why; i.e. what the spatial maps and time-series of those two components look like.

Again, I don't think the ICA decomposition is different among those conditions, but the classification is. That's why veryfing that the 67 components we get are the same across Python versions, hardware and CLI/workflow is the first step we should be taking.

**Comment by emdupre** [comment]
> Differences in hardware are part of the problem. Although github-actions uses the same software, the linux builds can use two different family of CPUs:

    Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
    Intel(R) Xeon(R) Platinum 8171M or 8272CL CPU @ 2.60GHz

We've [seen this in the past](https://github.com/mne-tools/mne-python/issues/4922)(that it's lower level features like BLAS / LAPACK version that are driving ICA differences), but that had been patched previously in scikit-learn, which we're using here. I don't know if something similar (though new) could be driving this, though.

I can check on the bug reports in some of our dependencies, but I agree with @eurunuela that it'd be helpful to know if you're getting the same ICA components with different classifications _or_ if you're actually getting different ICA components on the different architectures. You should be able to just compare `pca_mixing.tsv` and `ica_mixing.tsv` between two different runs.

**Comment by notZaki** [comment]
I suppose not much can be done about hardware differences, but I tried looking into the difference between using CLI/Workflow because as @eurunuela noted, it doesn't make sense---they should be calling identical code. The simplest answer would be that I'm making a mistake when using those two methods.

Here's the test functions that I'm running---it's adapted from the [existing three-echo test](https://github.com/ME-ICA/tedana/blob/45e5ad7ab23f89dad091c3f0b6c7cc61062a0acc/tedana/tests/test_integration.py#L155) and there's some unused variables in my version, but I think both tests should give the same output:
```python
def test_integration_three_echo_debug_cli(skip_integration):
    """ Integration test of the full tedana workflow using three-echo test data
    """

    if skip_integration:
        pytest.skip('Skipping three-echo integration test')
    out_dir = '/tmp/data/three-echo/TED.three-echo'
    if os.path.exists(out_dir):
        shutil.rmtree(out_dir)

    # download data and run the test
    download_test_data('https://osf.io/rqhfc/download',
                       os.path.dirname(out_dir))

    args = (['-d', '/tmp/data/three-echo/three_echo_Cornell_zcat.nii.gz',
             '-e', '14.5', '38.5', '62.5',
             '--tedpca', 'mdl',
             '--out-dir', out_dir, '--debug', '--verbose'])
    tedana_cli._main(args)


def test_integration_three_echo_debug_workflow(skip_integration):
    """ Integration test of the full tedana workflow using three-echo test data
    """

    if skip_integration:
        pytest.skip('Skipping three-echo integration test')
    out_dir = '/tmp/data/three-echo/TED.three-echo'
    out_dir2 = '/tmp/data/three-echo/TED.three-echo-rerun'
    if os.path.exists(out_dir2):
        shutil.rmtree(out_dir2)

    # download data and run the test
    download_test_data('https://osf.io/rqhfc/download',
                       os.path.dirname(out_dir))
    tedana_cli.tedana_workflow(
        data='/tmp/data/three-echo/three_echo_Cornell_zcat.nii.gz',
        tes=[14.5, 38.5, 62.5],
        out_dir=out_dir2,
        debug=True,
        verbose=True,
        tedpca='mdl')
```

The outputs for the two tests above is [on my fork](https://github.com/notZaki/tedana/tree/6c6c7fb4fdea74d2aee071761b298dc1138f5a56/data/three-echo)---the `*rerun` folder uses `tedana_workflow` while the non-suffix one uses `_main` (CLI)---obtained from a github-actions instance running MacOS [[log]](https://github.com/notZaki/tedana/runs/1519451059?check_suite_focus=true)

The values in `pca_mixing.tsv` are similar between the two runs---the maximum difference between them is on the order of 1e-13. Could be because of floating point errors, but not sure if it should be the case if the same code is being run.

The values in `ica_mixing.tsv` were not similar---the maximum difference between them is ~5. I thought the order of components might've changed, so I tried sorting them by their sum before taking the difference, but they still didn't match.

Direct link to `ica_mixing.tsv`:

- [Using CLI](https://github.com/notZaki/tedana/blob/6c6c7fb4fdea74d2aee071761b298dc1138f5a56/data/three-echo/TED.three-echo/ica_mixing.tsv)
- [Using Workflow](https://github.com/notZaki/tedana/blob/6c6c7fb4fdea74d2aee071761b298dc1138f5a56/data/three-echo/TED.three-echo-rerun/ica_mixing.tsv)

**Comment by eurunuela** [comment]
It looks like you're running the same command on both tests, so that's good.

It makes sense that there are no differences on the PCA (you always get the same number of selected components, which is what we use the PCA for) and the differences you see are definitely a result of floating-point errors.

I've checked the `ica_mixing.tsv` files you linked in your previous comment and it looks like it is an ICA issue, meaning that the ICA components are different between them. This is a big problem. I calculated the correlation of the components to see this:

```
[ 0.99 -0.08 -0.09  0.26  0.41  0.98 -0.03  0.15  0.19 -0.08 -0.06  0.88
  0.    0.05  0.05  0.    0.01 -0.51  0.04  0.2   0.13 -0.12  0.05 -0.11
  0.35 -0.07  0.07  0.15 -0.02  0.47  0.05  0.93 -0.06  0.02  0.01 -0.01
 -0.02 -0.12 -0.07 -0.26 -0.1   0.06  0.03  0.   -0.21  0.11  0.03 -0.15
  0.09 -0.18 -0.07  0.03  0.06 -0.   -0.39  0.65 -0.41 -0.15  0.12 -0.08
 -0.08 -0.11 -0.3  -0.13 -0.04 -0.18  0.12  0.21] 
```

So, as you can see, a few of them are highly correlated, like component 00.

![Figure_1](https://user-images.githubusercontent.com/13706448/101611281-9b53a900-3a09-11eb-8a3b-8c73a717bb38.png)

But most of them are completely different, like component 12.

![Figure_2](https://user-images.githubusercontent.com/13706448/101611321-a73f6b00-3a09-11eb-9e18-4fce890d9029.png)

I guess the next step should be to see if the differences are a result of the way you're calling tedana (cli vs workflow). So, I would rerun the same test on the same conditions (same hardware and Python version) and see if we get the same results.

**Comment by tsalo** [comment]
> I've checked the ica_mixing.tsv files you linked in your previous comment and it looks like it is an ICA issue, meaning that the ICA components are different between them. This is a big problem.

Is this a problem, though? ICA is stochastic, so if anything differed across approaches _that_ is where I would expect the differences to come from. I would be far more worried if the differences came from the metrics or decision tree (which we still need to check, right?).

**Comment by jbteves** [comment]
Well, I think the issue is that theoretically the approaches should be nearly identical. We're talking about pretty minor platform variations resulting in entirely different components, which I don't think should happen. I would hope that with the same seed on the same operating system they would at least be comparable components, rather than completely different.

**Comment by tsalo** [comment]
With an approach like ICA, I think that anything short of exactly the same should be as different as anything else, if that makes sense. I could be wrong though.

**Comment by eurunuela** [comment]
I agree with you @tsalo . However, we're using [fastICA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FastICA.html), which should yield reproducible results when using a seed. The results @notZaki is getting are far from reproducible.

**Comment by notZaki** [comment]
The effect of software/hardware on the decomposition was understandable but I didn't understand why `tedana_cli.tedana_workflow` and `tedana_cli._main` were giving different results.

Going through the thread [mentioned by](https://github.com/ME-ICA/tedana/issues/629#issuecomment-740805111) @emdupre, one of the discussion points was that the number of threads affects the decomposition. This is likely why workflow/CLI are giving different results---calling `tedana_cli.tedana_workflow` directly will use all available threads while `tedana_cli._main` will default to a single thread:
https://github.com/ME-ICA/tedana/blob/45e5ad7ab23f89dad091c3f0b6c7cc61062a0acc/tedana/workflows/tedana.py#L674-L681

This was confirmed by re-running the tests on github-actions with number of threads restricted to 1---there was no longer any difference between using workflow or CLI. [[logs]](https://github.com/notZaki/tedana/actions/runs/442528188). This also explains the gif in the original post---I was using `tedana_workflow` on a shared server, and was probably allocated a different number of threads in the two runs.

So the difference between workflow/CLI can also be explained by differences in hardware (number of threads). 

---

I don't think there is any single way to fix the inconsistent decomposition. Perhaps additional preprocessing (e.g. ZCA) might make the decomposition more stable, or maybe the current situation is due to poor data quality. 

I'm going to close this issue because I think the underlying causes have been identified.
